runtime-native/src/lib.rs

```diff
@@ -10,64 +10,12 @@
     rust_2018_idioms
 )]
 
-use futures::prelude::*;
-use futures::{future::BoxFuture, task::SpawnError};
-use lazy_static::lazy_static;
-
-use std::io;
-use std::net::SocketAddr;
-use std::pin::Pin;
-
-mod tcp;
-mod udp;
-
-use tcp::{TcpListener, TcpStream};
-use udp::UdpSocket;
-
-lazy_static! {
-    static ref JULIEX_THREADPOOL: juliex::ThreadPool = {
-        juliex::ThreadPool::with_setup(|| {
-            runtime_raw::set_runtime(&Native);
-        })
-    };
-}
-
-/// The Native runtime.
-#[derive(Debug)]
-pub struct Native;
-
-impl runtime_raw::Runtime for Native {
-    fn spawn_boxed(&self, fut: BoxFuture<'static, ()>) -> Result<(), SpawnError> {
-        JULIEX_THREADPOOL.spawn_boxed(fut.into());
-        Ok(())
-    }
-
-    fn connect_tcp_stream(
-        &self,
-        addr: &SocketAddr,
-    ) -> BoxFuture<'static, io::Result<Pin<Box<dyn runtime_raw::TcpStream>>>> {
-        let romio_connect = romio::TcpStream::connect(addr);
-        let connect = romio_connect.map(|res| {
-            res.map(|romio_stream| {
-                Box::pin(TcpStream { romio_stream }) as Pin<Box<dyn runtime_raw::TcpStream>>
-            })
-        });
-        connect.boxed()
-    }
-
-    fn bind_tcp_listener(
-        &self,
-        addr: &SocketAddr,
-    ) -> io::Result<Pin<Box<dyn runtime_raw::TcpListener>>> {
-        let romio_listener = romio::TcpListener::bind(&addr)?;
-        Ok(Box::pin(TcpListener { romio_listener }))
-    }
-
-    fn bind_udp_socket(
-        &self,
-        addr: &SocketAddr,
-    ) -> io::Result<Pin<Box<dyn runtime_raw::UdpSocket>>> {
-        let romio_socket = romio::UdpSocket::bind(&addr)?;
-        Ok(Box::pin(UdpSocket { romio_socket }))
-    }
-}
+#[cfg(target_arch = "wasm32")]
```

Comment:
Thinking about it more, I think it might be worth thinking about this as a feature instead of purely how it works on a specific target. The `wasm32-unknown-unknown` target literally assumes nothing other than that it's wasm32. By using wasm-bindgen we make the assumption on behalf of the user that they will run this in a browser. With WASI around the corner, this might cause confusion or might even be the wrong thing to do.

Reply:
If I'm understanding you correctly you're thinking this might break forwards compatibility with WASI, right?

I'm not necessarily convinced that's the case. The current conversation around async support in WASI seems to be centered around providing an execution model _inside_ the WASI runtime, rather than have a way to open sockets with an asynchronous flag. This is quite a departure from the `epoll` architecture that `mio` employs, meaning that whichever way we turn things, it won't work in WASI.

What I'm thinking we can do after this patch lands, is to further differentiate between the WASM targets by using more specific `cfg`s once the async WASI model further stabilizes.

Perhaps we could even pre-empt that and add a WASI target that panics on _every_ call.

Thoughts?

Reply:
It's not only WASI - what we're effectively saying is that by compiling for any wasm32-*-* target, we are requiring that it be run in the browser. 

I think making any kind of assumptions on how WASI will work is not a good idea. The spec is at the absolute beginning of its life. We can't really be sure of any direction it will take.

Reply:
FWIW, WASI is a different target from `wasm32-unknown-unknown`; it is `wasm32-unknown-wasi`.

Probably you want to do `cfg(all(feature = "wasm-bindgen", target_arch = "wasm32"))`, which should also be forward compatible with `wasm-bindgen` on WASI.

Reply:
Updated with @fitzgen's recommendation! :sparkles: